### ISSUE TITLE
Fix S3247: FP when casting on different identifier

### ISSUE THREAD

**Issue by ValentijnMakkenze** [SonarSource/sonar-dotnet#2314]
### Description
We get a false positive for Replace this type-check-and-cast sequence with an 'as' and a null check

### Repro steps

I simplified the code a lot but we have an if where we check the type and then do stuff where we cast a function result to the same type and that gives a warning on the if. 

```
    public void TestDuplicateCasts(object x)
    {
      if (x is S3247Test) // Replace this type-check-and-cast sequence with an 'as' and a null check
      {
        var b = new S3247Test();
        var c = (S3247Test)b;
      }
    }
```

### Expected behavior
No warning

### Actual behavior
Warning	S3247	Replace this type-check-and-cast sequence with an 'as' and a null check.

### Known workarounds
Mark item as false positive

### Related information
* SonarC# Version 7.11.0.8083
* Visual Studio Version 2017
* If running through the Scanner for MSBuild, its version 4.5.0.1761

Reproduced in VS2019 with SonarAnalyzer.CSharp 7.11.0.8083

**Comment by Evangelink** [comment]
Hi @ValentijnMakkenze,

Thanks for the feedback! That's an interesting kind of false positive, I'll have a look and keep in informed about my progress.
